File: frontend/src/api/rehearsal.ts
import axios from 'axios'
import { api } from './index'

export interface PageTiming {
  page_number: number
  start_sec: number
  end_sec: number
}

export interface FillerWordDetail {
  word: string
  count: number
  positions: number[]
}

export interface PageScore {
  page_number: number
  actual_duration_sec: number
  suggested_duration_sec: number
  timing_ok: boolean
  importance_level: number
}

export interface RehearsalStatus {
  rehearsal_id: number
  status: number
  total_score: number | null
  fluency_score: number | null
  rate_score: number | null
  timing_score: number | null
}

export interface RehearsalReport extends RehearsalStatus {
  pitch_task_id: number
  audio_url: string
  audio_duration: number
  filler_count: number | null
  filler_detail: FillerWordDetail[]
  chars_per_min: number | null
  total_duration_sec: number | null
  improvement_tips: string[]
  page_scores: PageScore[]
  page_timings: PageTiming[]
  created_at: string
}

export interface RehearsalListItem {
  rehearsal_id: number
  pitch_task_id: number
  status: number
  total_score: number | null
  audio_duration: number
  created_at: string
}

export const rehearsalApi = {
  start: (pitchTaskId: number, planId: number | null) =>
    api.post<{ rehearsal_id: number; upload_url: string; object_key: string }>(
      '/rehearsals/start',
      { pitch_task_id: pitchTaskId, plan_id: planId }
    ),

  uploadAudio: (uploadUrl: string, blob: Blob, onProgress?: (pct: number) => void) =>
    axios.put(uploadUrl, blob, {
      headers: { 'Content-Type': 'audio/webm' },
      onUploadProgress: (e) => {
        if (onProgress && e.total) onProgress(Math.round((e.loaded / e.total) * 100))
      },
    }),

  complete: (rehearsalId: number, data: { object_key: string; audio_duration: number; page_timings: PageTiming[] }) =>
    api.post<RehearsalStatus>(`/rehearsals/${rehearsalId}/complete`, data),

  getStatus: (rehearsalId: number) =>
    api.get<RehearsalStatus>(`/rehearsals/${rehearsalId}/status`),

  getReport: (rehearsalId: number) =>
    api.get<RehearsalReport>(`/rehearsals/${rehearsalId}/report`),

  listByTask: (pitchTaskId: number) =>
    api.get<RehearsalListItem[]>(`/rehearsals/task/${pitchTaskId}`),
}

File: backend/app/schemas/rehearsal.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field


class PageTiming(BaseModel):
    page_number: int
    start_sec: float
    end_sec: float


class StartRehearsalRequest(BaseModel):
    pitch_task_id: int
    plan_id: int | None = None


class StartRehearsalResponse(BaseModel):
    rehearsal_id: int
    upload_url: str     # presigned MinIO PUT URL
    object_key: str


class CompleteRehearsalRequest(BaseModel):
    object_key: str             # MinIO key of uploaded audio
    audio_duration: int         # seconds
    page_timings: list[PageTiming] = Field(default_factory=list)


class RehearsalStatusResponse(BaseModel):
    rehearsal_id: int
    status: int
    total_score: float | None = None
    fluency_score: float | None = None
    rate_score: float | None = None
    timing_score: float | None = None


class FillerWordDetail(BaseModel):
    word: str
    count: int
    positions: list[int]


class PageScore(BaseModel):
    page_number: int
    actual_duration_sec: float
    suggested_duration_sec: float
    timing_ok: bool
    importance_level: int


class RehearsalReportResponse(BaseModel):
    rehearsal_id: int
    pitch_task_id: int
    status: int
    audio_url: str
    audio_duration: int
    total_score: float | None = None
    fluency_score: float | None = None
    rate_score: float | None = None
    timing_score: float | None = None
    filler_count: int | None = None
    filler_detail: list[FillerWordDetail] = Field(default_factory=list)
    chars_per_min: float | None = None
    total_duration_sec: float | None = None
    improvement_tips: list[str] = Field(default_factory=list)
    page_scores: list[PageScore] = Field(default_factory=list)
    page_timings: list[PageTiming] = Field(default_factory=list)
    created_at: datetime


class RehearsalListItem(BaseModel):
    rehearsal_id: int
    pitch_task_id: int
    status: int
    total_score: float | None = None
    audio_duration: int
    created_at: datetime
